packages/fleather/lib/src/widgets/history.dart

```diff
@@ -0,0 +1,280 @@
+import 'dart:async';
+
+import 'package:fleather/fleather.dart';
+import 'package:flutter/widgets.dart';
+import 'package:quill_delta/quill_delta.dart';
+
+/// A void function that takes a [TextEditingValue].
+@visibleForTesting
+typedef TextEditingValueCallback = void Function(TextEditingValue value);
+
+/// Provides undo/redo capabilities for text editing.
+///
+/// Listens to [controller] as a [ValueNotifier] and saves relevant values for
+/// undoing/redoing. The cadence at which values are saved is a best
+/// approximation of the native behaviors of a hardware keyboard on Flutter's
+/// desktop platforms, as there are subtle differences between each of these
+/// platforms.
+///
+/// Listens to keyboard undo/redo shortcuts and send update to [controller].
+class FleatherHistory extends StatefulWidget {
+  /// Creates an instance of [FleatherHistory].
+  const FleatherHistory(
+      {super.key, required this.child, required this.controller});
+
+  /// The child widget of [FleatherHistory].
+  final Widget child;
+
+  /// The [FleatherController] to save the state of over time.
+  final FleatherController controller;
+
+  @override
+  State<FleatherHistory> createState() => _FleatherHistoryState();
+}
+
+class _FleatherHistoryState extends State<FleatherHistory> {
+  late final _Throttled<Delta> _throttledPush;
+  late HistoryStack _stack;
+  Timer? _throttleTimer;
+
+  // This duration was chosen as a best fit for the behavior of Mac, Linux,
+  // and Windows undo/redo state save durations, but it is not perfect for any
+  // of them.
+  static const Duration _kThrottleDuration = Duration(milliseconds: 500);
+
+  void _undo(UndoTextIntent intent) {
+    _update(_stack.undo());
+  }
+
+  void _redo(RedoTextIntent intent) {
+    _update(_stack.redo());
+  }
+
+  void _update(Delta? changeDelta) {
+    if (changeDelta == null || changeDelta.isEmpty) {
+      return;
+    }
+
+    widget.controller.compose(changeDelta,
+        selection: _fromDelta(changeDelta), source: ChangeSource.local);
+  }
+
+  TextSelection _fromDelta(Delta changeDelta) {
+    assert(changeDelta.isNotEmpty);
+    final firstOp = changeDelta.first;
+    int baseOffset = 0;
+    if (firstOp.isRetain && firstOp.attributes == null) {
+      baseOffset = firstOp.length;
+    }
+    int extentOffset = baseOffset;
+    final lastOp = changeDelta.last;
+    if (lastOp.isRetain && lastOp.attributes != null) {
+      extentOffset = changeDelta.textLength;
+    }
+    if (lastOp.isInsert) {
+      baseOffset = changeDelta.textLength;
```

Comment:
add a comment?
explain why we would like the cursor to be at the beginning of the selection ?